### PR TITLE
Implement Distributed tracing using OpenTelemtry

### LINKED ISSUE #1469: Add OpenTracing/Census/Telemetry support for OPA API's
Some history on the drive for this in #1451 

Basically we want to allow for systems to do tracing of requests that flow through OPA for better system-wide visibility.

As-is a user of OPA can pass in arbitrary inputs and via the policy/query arbitrary outputs for external request/trace ID's. OPA will also generate a unique decision ID as it does evaluations.

It would be ideal if we supported using/forwarding trace contexts via some standard spec like OpenTracing, OpenCensus, in the future OpenTelemetry, etc. On its own this can give some powerful debug/tracing features for OPA internally, but even more powerful when integrated with larger systems that are utilizing these tracing frameworks.

The changes required would be to use something like the OpenTracing golang library in OPA (we could maybe avoid vendoring something too, not sure how much code we really need). Then on the API server we need to extract the span context, or make a new one if it wasn't specified. Then pass it around with the context we already use everywhere internally. We could go crazy with spans and tracing stuff in OPA but as a starting point just having entry/exit to the REST API is probably enough for most use-cases.

Reference:
https://opentelemetry.io/
https://opentracing.io/
https://opentracing.io/guides/golang/inject-extract/
https://opentracing.io/guides/golang/

### PR DESCRIPTION
**Pull request by rvalkenaers**
This commit implements tracing using the net/http [automatic instrumentation wrappers](https://opentelemetry.io/docs/go/instrumentation/#automatic-instrumentation) on the server and topdown/http packages.

Following configuration flags are added:
      --distributed-tracing                       enable distributed tracing using OpenTelemetry Tracing
      --distributed-tracing-address string        address of the OpenTelemetry Collector gRPC endpoint (default "localhost:4317")
      --distributed-tracing-sample-rate int       precentage of traces that are sampled and exported (default 100)
      --distributed-tracing-service-name string   logical name of the service used by OpenTelemetry Tracing (default "opa")
      --distributed-tracing-tls {off,tls,mtls}            set distributed tracing tls scheme (default off)
      --distributed-tracing-tls-cert-file string          set path of TLS certificate file of the tracing exporter
      --distributed-tracing-tls-private-key-file string   set path of TLS private key file of the tracing exporter
      --distributed-tracing-tls-skip-verify               disable certificate chain verification
Fixes #1469
Signed-off-by: Rien Valkenaers <rien.valkenaers@gmail.com>

This is an example trace for a "[Pull Data during Evaluation](https://www.openpolicyagent.org/docs/latest/external-data/#option-5-pull-data-during-evaluation)" scenario:
opa_trace

Screenshot 2021-11-17 at 17 03 56

### CODE REVIEW

File: distributedtracing/distributedtracing.go
Comment (srenatus) [inline distributedtracing/distributedtracing.go:21]: Is this what everyone does? I could imagine wanting to have a TLS option here 🤔

File: distributedtracing/distributedtracing.go
Comment (srenatus) [inline distributedtracing/distributedtracing.go:16]: Hmm package-level vars for configuration feel a bit like a shortcut 🤔 Is the alternative unbearable? I can't tell without trying myself.

**Review by srenatus — COMMENTED** [review]
Thanks for pushing this. It'll be great to have tracing support soon. 🎉 

Two things that caught my eye on a quick glance, please see comments inline 👇 🙃

File: distributedtracing/distributedtracing.go
Comment (rvalkenaers) [reply distributedtracing/distributedtracing.go:16]: I couldn't find an easy way to access it from https://github.com/open-policy-agent/opa/blob/main/topdown/http.go.
Any suggestions are more than welcome :slightly_smiling_face:

Comment (srenatus) [reply distributedtracing/distributedtracing.go:16]: We'd thread it through builtincontext, I think.

Comment (bhoriuchi) [reply distributedtracing/distributedtracing.go:16]: generally what i have done if i want to conditionally enable tracing is default to the noop tracer https://github.com/open-telemetry/opentelemetry-go/blob/main/trace/noop.go#L27 if tracing configuration is not present then there is no need to add logic on whether to instrument tracing or not, everything just gets instrumented.

File: distributedtracing/distributedtracing.go
Comment (rvalkenaers) [reply distributedtracing/distributedtracing.go]: Added following flags to support tls
```
--distributed-tracing-tls {off,tls,mtls}            set distributed tracing tls scheme (default off)
--distributed-tracing-tls-cert-file string          set path of TLS certificate file of the tracing exporter
--distributed-tracing-tls-private-key-file string   set path of TLS private key file of the tracing exporter
--distributed-tracing-tls-skip-verify               disable certificate chain verification
```

**Comment by anderseknert** [comment]
Love this! 👍 

One thing I'm not too sure about though is all the new arguments added to `opa run`. Not in the sense that they aren't necessary, but I wonder if it'd be a good idea to instead have them added to the OPA configuration? In addition to adding clutter to the `opa run --help` output, command line arguments have some disadvantages over config, including:

1. These attributes aren't available for review at the `/v1/config` endpoint.
2. They can't be set/modified by discovery bundles.
3. They can't be set programmatically / by the SDK (not sure if applicable here though).

We've already seen this being problematic [elsewhere](https://github.com/open-policy-agent/opa/issues/3980), so perhaps this would be a good candidate for config rather than CLI arguments? Note that it's still possible to set config attributes through the CLI with `--set config=value` if one prefers that.

Thoughts, @srenatus @tsandall ?

**Comment by srenatus** [comment]
@anderseknert I'm on board with making this part of the config instead :+1: Thanks for compiling the pros and cons.

File: distributedtracing/distributedtracing.go
Comment (srenatus) [reply distributedtracing/distributedtracing.go:20]: Anything the builtin `http.send` needs from the evaluator would be passed down via the [`BuiltinContext`](https://github.com/open-policy-agent/opa/blob/v0.34.2/topdown/builtins.go#L34-L53). So we could add fields there.

The question I can't answer right now is what we'd want to have there: 
- [`otelhttp.Option`](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp#Option) I think if we include that, it'll be enough to have all the tracer config stuff passed along?
- Do we care for anything evaluation-related? Would this be a descendent of some span that corresponds to the incoming request...? If so, how is that information passed around?

Sorry if this is vague, I really need you input here. Just trying to wrap my mind around how it would work and be most useful.

**Review by srenatus — COMMENTED** [review]
Two comments about what "keeping the embeddings unaffected" means for this PR. What do you think?

File: distributedtracing/distributedtracing.go
Comment (rvalkenaers) [reply distributedtracing/distributedtracing.go]: Pass []otelhttp.Option along in stead of using global trace provider

File: cmd/run.go
Comment (srenatus) [inline cmd/run.go:216]: As @anderseknert mentioned, it would be nice to put these into the yaml/json config instead. I'm trying to find a few pointers on what to do where 👀

**Review by srenatus — COMMENTED** [review]
The setup without the package-level global looks good to me.

I believe there are a few points we should still get figured out:

1. Can we have some sort of e2e test? There're a bunch of them in `test/e2e`, and I would imagine that some sort of mock (or in-memory) trace collector could be used to assert that the right things get sent for requests.
2. The config via json/yaml instead of CLI args would be nice. I'll try to find an example of how we do it in other places...

I've seen the additions to the v1 data API, would you consider that a first step, and that the other APIs would eventually follow, or is this something you'd perhaps want to include in this PR? I'd thing that the v0 data API could use this, too, and perhaps also the policy API. What do you think? If it's getting unwieldy, I suppose we could tackle that in follow-up PRs, too.

**Comment by srenatus** [comment]
@rvalkenaers thanks for bearing with me! This is a tremendous feature, but also quite a bit of work, I'm afraid 😅

**Comment by rvalkenaers** [comment]
@srenatus 

> I've seen the additions to the v1 data API, would you consider that a first step, and that the other APIs would eventually follow, or is this something you'd perhaps want to include in this PR? I'd thing that the v0 data API could use this, too, and perhaps also the policy API. What do you think? If it's getting unwieldy, I suppose we could tackle that in follow-up PRs, too.

I've did a quick test and it's actually creating spans for all API endpoints.

**Comment by srenatus** [comment]
> I did a quick test and we're actually creating spans for all API endpoints.

Also the v0 data API with http.send and all that? I thought we hadn't wired that config into the call to rego.New.

**Comment by rvalkenaers** [comment]
> Also the v0 data API with http.send and all that? I thought we hadn't wired that config into the call to rego.New.

For the v0 data API we only have the server spans, not the http.send spans. I would prefer to implement those in follow-up PRs.

**Comment by srenatus** [comment]
@rvalkenaers I'm not actually completely sure how the config change would look like -- @anderseknert, would making the config a piece of JSON mean that _this_ would be a plugin? Like status, logs, bundles...

**Comment by anderseknert** [comment]
I hadn't thought about that. I guess it makes sense from an architectural POV, but you're probably better equipped to judge that :) On the other hand, if this attaches to both the server component and things like http.send, maybe treating it like an integrated component makes things simpler..

**Comment by srenatus** [comment]
> On the other hand, if this attaches to both the server component and things like http.send, maybe treating it like an integrated component makes things simpler..

Yup I think that's the case. I think we'll need something along these lines, then, right? 👉 https://github.com/open-policy-agent/opa/blob/v0.35.0/topdown/cache/cache.go#L22-L60 Wired up in the same way, it would be configurable either via CLI `--set` flags, or via a discovery bundle.

Let's also ensure that this gets documented so that people are aware of this cool new feature. But we should probably first settle on _how_ to configure it.

**Comment by srenatus** [comment]
Sorry for the wait, I haven't forgotten about this 😅 I'll play with this some locally, but it looks good!

File: docs/content/configuration.md
Comment (srenatus) [inline docs/content/configuration.md:78]: 🤦 It's not you, it's YAML 1.11 which is used with the config file parser.... anyways, this config will fail to load:

```
error: config error: json: cannot unmarshal bool into Go struct field distributedTracingConfig.encryption of type string 
```
because `off` is interpreted as `false`. To fix the example, we need quotes:
```suggestion
  encryption: "off"
```

**Review by srenatus — COMMENTED** [review]
I've stumbled upon something we'll need to fix here: logging.

When I run this locally, with DT enabled, and the endpoint isn't available, I see a bunch of these:

```
{"client_addr":"127.0.0.1:50944","level":"info","msg":"Received request.","req_id":7,"req_method":"GET","req_path":"/health","time":"2021-12-09T13:34:23+01:00"}
{"client_addr":"127.0.0.1:50944","level":"info","msg":"Sent response.","req_id":7,"req_method":"GET","req_path":"/health","resp_bytes":2,"resp_duration":0.291205,"resp_status":200,"time":"2021-12-09T13:34:23+01:00"}
2021/12/09 13:34:27 context deadline exceeded
2021/12/09 13:34:27 traces exporter is disconnected from the server 127.0.0.1:4317: context deadline exceeded
2021/12/09 13:34:42 context deadline exceeded
{"client_addr":"127.0.0.1:50981","level":"info","msg":"Received request.","req_id":8,"req_method":"GET","req_path":"/health","t
```

It's a mix of OPA's logs, using its logging infrastructure, and some other logs, coming from the DT client. We somehow need to get the latter logs to use OPA's logging, too.

I haven't dug in deep yet to find a fix, but maybe you know what to do right away? 😃

**Comment by srenatus** [comment]
👉 https://pkg.go.dev/go.opentelemetry.io/otel#SetErrorHandler could using this be the way to go?

> The default ErrorHandler instance returned will log all errors to STDERR until an override ErrorHandler is set with SetErrorHandler. All ErrorHandler returned prior to this will automatically forward errors to the set instance instead of logging.

**Comment by rvalkenaers** [comment]
> 👉 https://pkg.go.dev/go.opentelemetry.io/otel#SetErrorHandler could using this be the way to go?
> 
> > The default ErrorHandler instance returned will log all errors to STDERR until an override ErrorHandler is set with SetErrorHandler. All ErrorHandler returned prior to this will automatically forward errors to the set instance instead of logging.

I had this in place initially (https://github.com/open-policy-agent/opa/blob/e93a88d7c4c74abde5d8546f79b31b171afa9ad4/distributedtracing/distributedtracing.go#L45)

I removed it when I got rid of the global vars but I didn't found another way to do it.

**Comment by srenatus** [comment]
Ahh, alright. Are there non-global ways to provide an error handler?

File: runtime/runtime.go
Comment (srenatus) [inline runtime/runtime.go:327]: [nit] Let's do
```suggestion
		return nil, fmt.Errorf("config error: %w", err)
```

**Comment by srenatus** [comment]
Can we put the call to set up the global handler in a place where it won't affect people embedding OPA? I'm thinking about `cmd/run.go`... That way, if you use the SDK, start a runtime, etc, it'll all not be affected; but if you use `opa run -s`, you'll get the proper logging setup. What do you think?

**Comment by rvalkenaers** [comment]
> Can we put the call to set up the global handler in a place where it won't affect people embedding OPA? I'm thinking about `cmd/run.go`... That way, if you use the SDK, start a runtime, etc, it'll all not be affected; but if you use `opa run -s`, you'll get the proper logging setup. What do you think?

Sounds good to me.

**Comment by srenatus** [comment]
Thank you, logging looks alright now:

```
{"client_addr":"127.0.0.1:53744","level":"info","msg":"Received request.","req_id":12,"req_method":"GET","req_path":"/health","time":"2021-12-10T11:17:56+01:00"}
{"client_addr":"127.0.0.1:53744","level":"info","msg":"Sent response.","req_id":12,"req_method":"GET","req_path":"/health","resp_bytes":2,"resp_duration":0.280679,"resp_status":200,"time":"2021-12-10T11:17:56+01:00"}
{"level":"error","msg":"Distributed tracing: context deadline exceeded","time":"2021-12-10T11:18:04+01:00"}
{"level":"error","msg":"Distributed tracing: max retry time elapsed: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:4317: connect: connection refused\"","time":"2021-12-10T11:18:09+01:00"}
```

File: internal/distributedtracing/distributedtracing.go
Comment (srenatus) [inline internal/distributedtracing/distributedtracing.go:31]: I'm not sure I understand why we'd need these. We could also _not_ set them "as defaults" in `validateAndInjectDefaults`...

File: internal/distributedtracing/distributedtracing.go
Comment (srenatus) [inline internal/distributedtracing/distributedtracing.go:58]: For string values, we should only need `*string` when we **need to differentiate** `""` and _not set_. Is that the case here? At least for TLS, I doubt it -- empty string isn't a valid default.

**Review by srenatus — COMMENTED** [review]
A couple of small-impact nitpicks, I think we're good to go then. 🚀 Thanks for bearing with me!

File: internal/distributedtracing/distributedtracing.go
Comment (rvalkenaers) [reply internal/distributedtracing/distributedtracing.go:58]: Good point, empty string shouldn't be valid for any of these

**Review by tsandall — COMMENTED** [review]
@rvalkenaers first of all, thank you SO much for putting in the work to enable tracing in OPA, it's much appreciated :pray:

I agree with @srenatus that this is very close to being mergeable--there's just one thing I'd like us to discuss before merging and that has to do w/ package dependencies:

1. The rego and topdown packages have been extended to include the internal/distributedtracing#Options type in exported functions. The problem with that is that callers outside the top-level OPA package will not be able to provide values of this type, because the type is behind an internal package in OPA. I think we'll need to create a new top-level package (e.g., github.com/open-policy-agent/opa/tracing) where we export the necessary types... I think Options is the only type but I could be wrong.

2. These changes put a direct dependency on OpenTelemetry into OPA, and not just OPA but also the low-level packages (rego and topdown) that users rely on solely for policy evaluation. Today those packages are constructed intentionally to avoid large 3rdparty dependencies. I'd like to know how feasible it would be to define a small shim interface in OPA--this would go into the exported package I mentioned in (1). OPA would need to implement some boilerplate but it would remove the hard dependency on opentelemetry from the rego and topdown packages.

@srenatus PTAL and let me know if this makes sense.

**Comment by srenatus** [comment]
@tsandall I agree, the dependency thing is a good point. As far as I am concerned, I'd leave this up to @rvalkenaers: I could take on dealing with those two points between merge and 0.36.0 (i.e. next week); or you (@rvalkenaers) could deal with it before we merge. Either way is fine with me. 🔀 😄

**Comment by rvalkenaers** [comment]
> @tsandall I agree, the dependency thing is a good point. As far as I am concerned, I'd leave this up to @rvalkenaers: I could take on dealing with those two points between merge and 0.36.0 (i.e. next week); or you (@rvalkenaers) could deal with it before we merge. Either way is fine with me. 🔀 😄

@srenatus If you could pick it up it would be much appreciated :pray:

**Comment by srenatus** [comment]
@rvalkenaers OK sure! Would you mind squashing these commits, adding a short commit message, so I can push the merge button?

**Review by srenatus — APPROVED** [review]
Thanks a lot for your contribution that get's the ball rolling with OPA and OpenTelemetry 👏 🎉 🥳 

I'll merge this, and take on the remaining little cleanup we've discussed in another PR.

**Comment by anderseknert** [comment]
This is great! Thanks @rvalkenaers 👍

**Pull request merged**
